File: src/lib/req.js
import axios from 'axios';
import { settingsDB } from '../db/db';
import { invoke } from '@tauri-apps/api';

/**
 * @typedef {Object} CustomHeaders
 * @property {string} Miru-Url - Custom URL header
 * @property {string} [anyOtherHeader] - Other headers can also be included
 */

/**
 * @typedef {Object} CustomRequestOptions
 * @property {CustomHeaders} headers - The headers to include in the request
 * @property {string} [method] - The method for the request (e.g., GET, POST)
 */

/**
 * Sends a request to the specified URL with the given options and returns the response data.
 *
 * @param {string} url - The URL to send the request to.
 * @param {CustomRequestOptions} options - The options for the request.
 * @return {Promise<any>} A promise that resolves with the response data if the request is successful,
 *                   or rejects with an error if the request fails.
 */
export async function request(url, options) {
	let proxy = await settingsDB.getSettings('proxy');
	url = proxy?.value + url;
    let response;
	try {
		if (window.__TAURI__) {	
			response = await invoke('make_request', { url, options });
		} else {
			response = (await axios(url, options)).data;
		}

		if (response.status === 200) {
			return response;
		} else {
			throw new Error('Failed to fetch manga list');
		}
	} catch (error) {
		throw new Error('Failed to fetch manga list');
	}
}

File: src-tauri/src/req.rs
use reqwest::Client;
use reqwest::Method;
use std::collections::HashMap;

#[derive(Debug, serde::Deserialize)] // Add this line
pub struct CustomRequestOptions {
    pub headers: HashMap<String, String>,
    pub method: Option<Method>,
}
pub async fn request(
    url: &str,
    options: CustomRequestOptions,
) -> Result<String, Box<dyn std::error::Error>> {
    let client = Client::new();

    let mut req = client.request(options.method.unwrap_or(reqwest::Method::GET), url);

    for (key, value) in options.headers {
        req = req.header(key, value);
    }

    let response = req.send().await?;

    if response.status().is_success() {
        let text = response.text().await?;
        Ok(text)
    } else {
        let error_message = format!("Failed to fetch data: {}", response.status());
        Err(error_message.into())
    }
}
